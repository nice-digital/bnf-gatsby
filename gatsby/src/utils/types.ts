import { type Primitive, type Merge, type Simplify } from "type-fest";

export type BuiltIns = Primitive | Date | RegExp;

export interface RecordSection extends Slug {
	order: number;
	title: string;
	content: string;
}

export interface MenuPageLink {
	href: string;
	title: string;
}

/** Represents the structure of JSON files used for meta descriptions:
 * a map of URL slug to string descriptions for BNF and BNFC */
export type MetaDescriptionsMap = Record<
	string,
	{ bnf: string | null; bnfc: string | null } | undefined
>;

export interface Slug {
	slug: string;
}

export interface SlugAndTitle extends Slug {
	title: string;
}

export type WithSlug<T extends object> = T & Slug;

/**
 * Add a `slug: string` property to each of the given `SlugTypes` recursively within the given type `Base`.
 *
 * This type is useful for augmenting raw feed types with `slug` properties when they're queried via GraphQl.
 * This is because we add slugs using the `@slug` directive in the GraphQL schema, so there are no slugs on the raw feed types.
 * It avoids having to do nested union types to add a `slug` property in each file that queries GraphQL.
 *
 * This type will often be used in tandem with `QueryResult` in practice to correctly type the results of GraphQL queries _and_ to add slugs where needed.
 *
 * 	@example To add slugs to all `FeedMedicinalForm` types within `FeedMedicinalForms`:
 * 		type MedicinalFormsWithSlugs = WithSlugDeep<FeedMedicinalForms, FeedMedicinalForm>
 * */
export type WithSlugDeep<
	Base extends object,
	SlugTypes extends object
> = Simplify<
	Merge<
		Base extends SlugTypes ? WithSlug<Base> : Base,
		{
			[Key in keyof Base]: Base[Key] extends (infer U)[] | undefined
				? U extends object
					? WithSlugDeep<U, SlugTypes>[] | undefined
					: Base[Key]
				: Base[Key] extends object
				? WithSlugDeep<Base[Key], SlugTypes>
				: Base[Key] extends object | undefined
				? WithSlugDeep<NonNullable<Base[Key]>, SlugTypes> | undefined
				: Base[Key];
		}
	>
>;

/**
 * Wrapper type for types returned from GraphQL queries.
 *
 * It does the following:
 * - replaces `undefined` with `null` (GraphQL queries return `null` for empty properties)
 * - makes arrays required (not `null` or `undefined`)
 * */
export type QueryResult<Base> = Base extends NonNullable<BuiltIns>
	? Base
	: Base extends BuiltIns
	? NonNullable<Base> | null
	: Base extends (infer U)[]
	? NonNullable<QueryResult<U>[]>
	: Base extends object
	? {
			[Key in keyof Base]-?: NonNullable<Base[Key]> extends unknown[]
				? NonNullable<QueryResult<Base[Key]>>
				: QueryResult<Base[Key]>;
	  }
	: unknown;
