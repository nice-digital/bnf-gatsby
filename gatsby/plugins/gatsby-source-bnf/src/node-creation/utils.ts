import type { NodeInput, SourceNodesArgs } from "gatsby";

export type BnfNodeType = `Bnf${string}`;

export type TypedNodeInput<
	TNodeType extends BnfNodeType,
	TNodeContent
> = NodeInput &
	TNodeContent & {
		// Internal Gatsby node stuff
		internal: {
			type: TNodeType;
		} & NodeInput["internal"];
	};

/**
 * Helper method to create a BNF-specific Gatsby GraphQL Node, that is, a node with a type starting with `Bnf`.
 *
 * @param nodeContent The content of the node to create, without any of the Gatsby internals
 * @param type The name of the node type, starting with `Bnf` e.g. `BnfDrug`
 * @param sourceNodesArgs The arguments to `sourceNodes` from Gatsby
 * @template TNodeContent The type of the node content
 */
export const createBnfNode = <TNodeContent extends { id: string }>(
	nodeContent: TNodeContent,
	type: BnfNodeType,
	{ createContentDigest, actions: { createNode } }: SourceNodesArgs
): void => {
	const content = JSON.stringify(nodeContent),
		contentDigest = createContentDigest(content);

	const node: TypedNodeInput<typeof type, TNodeContent> = {
		...nodeContent,
		internal: {
			type,
			content,
			contentDigest,
		},
	};

	createNode(node);
};
