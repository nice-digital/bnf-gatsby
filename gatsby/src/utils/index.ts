export * from "./types";

/**
 * Returns whether the given object is truthy or not but casted to the object's type.
 *
 * This casting allows filtering of types that are potentially falsy:
 *
 * @example
 * 	const list: (SomeType | undefined)[] = []
 * 	someList.filter(isTruthy).map(a: SomeType => {
 *  		// Do something
 *  	})
 * */
export const isTruthy = <TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe =>
	!!maybeT;
