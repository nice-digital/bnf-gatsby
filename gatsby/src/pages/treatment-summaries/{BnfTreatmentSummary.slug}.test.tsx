import { render, waitFor, screen, within } from "@testing-library/react";

import { SlugAndTitle } from "@/utils";

import TreatmentSummaryPage, {
	query,
	type TreatmentSummaryPageProps,
} from "./{BnfTreatmentSummary.slug}";

const treatmentSummary: TreatmentSummaryPageProps["data"]["bnfTreatmentSummary"] =
	{
		title: "Acne",
		relatedDrugs: [],
		relatedTreatmentSummaries: [],
		sections: [
			{
				title: "Section 1",
				slug: "section-1",
				content: "<p>Content for section 1</p>",
			},
			{
				title: "Section 2",
				slug: "section-2",
				content: "<p>Content for section 2</p>",
			},
		],
	};

const minimumProps: TreatmentSummaryPageProps = {
	data: {
		bnfTreatmentSummary: treatmentSummary,
	},
};

describe("TreatmentSummaryPage", () => {
	it("should match snapshot for graphql query", () => {
		expect(query).toMatchSnapshot();
	});

	describe("SEO", () => {
		it("should set page title from treatment summary title", async () => {
			render(<TreatmentSummaryPage {...minimumProps} />);

			await waitFor(() => {
				expect(document.title).toStartWith("Acne | Treatment summaries |");
			});
		});

		it("should set templated meta description", async () => {
			render(<TreatmentSummaryPage {...minimumProps} />);

			await waitFor(() => {
				expect(
					document
						// eslint-disable-next-line testing-library/no-node-access
						.querySelector("meta[name='description']")
				).toHaveAttribute(
					"content",
					"This treatment summary topic describes Acne"
				);
			});
		});
	});

	describe("Page header", () => {
		it("should add content start skip link target id to page header", () => {
			render(<TreatmentSummaryPage {...minimumProps} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			// eslint-disable-next-line testing-library/no-node-access
			expect(heading1.parentElement).toHaveAttribute("id", "content-start");
		});

		it("should render heading 1 with current page title", () => {
			render(<TreatmentSummaryPage {...minimumProps} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			expect(heading1).toHaveTextContent("Acne");
		});
	});

	describe("Breadcrumbs", () => {
		it.each([
			["NICE", "https://www.nice.org.uk/"],
			["BNF", "/"],
			["Treatment summaries", "/treatment-summaries/"],
		])(
			"should render default '(%s)' breadcrumb",
			(breadcrumbText, expectedHref) => {
				render(<TreatmentSummaryPage {...minimumProps} />);

				const breadcrumbNav = screen.getByRole("navigation", {
					name: "Breadcrumbs",
				});
				const breadcrumb = within(breadcrumbNav).getByRole("link", {
					name: breadcrumbText,
				});

				expect(breadcrumb).toHaveAttribute("href", expectedHref);
			}
		);

		it("should render current page breadcrumb without link", () => {
			render(<TreatmentSummaryPage {...minimumProps} />);

			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const currentPageCrumb = within(breadcrumbNav).getByText("Acne");
			expect(currentPageCrumb).toBeInTheDocument();
			expect(currentPageCrumb.tagName).toBe("SPAN");
		});
	});

	describe("body", () => {
		it("should render hash link to each record section", () => {
			render(<TreatmentSummaryPage {...minimumProps} />);
			expect(screen.getByRole("link", { name: "Section 1" })).toHaveAttribute(
				"href",
				"#section-1"
			);
			expect(screen.getByRole("link", { name: "Section 2" })).toHaveAttribute(
				"href",
				"#section-2"
			);
		});

		it("should render section with accessible name for each record section", () => {
			render(<TreatmentSummaryPage {...minimumProps} />);
			expect(
				screen.getByRole("region", { name: "Section 1" })
			).toBeInTheDocument();
			expect(
				screen.getByRole("region", { name: "Section 2" })
			).toBeInTheDocument();
		});

		describe.each([
			["Related drugs", "relatedDrugs" as const, "related-drugs", "drugs"],
			[
				"Related treatment summaries",
				"relatedTreatmentSummaries" as const,
				"related-treatment-summaries",
				"treatment-summaries",
			],
		])("%s", (sectionName, propertyName, expectedId, expectedBasePath) => {
			describe(`No ${sectionName}`, () => {
				it(`should not render ${sectionName} hash link when there are no ${sectionName}`, () => {
					render(<TreatmentSummaryPage {...minimumProps} />);
					expect(screen.queryByRole("link", { name: sectionName })).toBeNull();
				});

				it(`should not render ${sectionName} secttion when there are no ${sectionName}`, () => {
					render(<TreatmentSummaryPage {...minimumProps} />);
					expect(
						screen.queryByRole("region", { name: sectionName })
					).toBeNull();
				});
			});

			describe(`${sectionName} with content`, () => {
				const relatedThings: SlugAndTitle[] = [
					{
						title: "Thing 1",
						slug: "thing-1",
					},
					{
						title: "Thing 2",
						slug: "thing-2",
					},
				];

				const props: TreatmentSummaryPageProps = {
					data: {
						bnfTreatmentSummary: {
							...treatmentSummary,
							[propertyName]: relatedThings,
						},
					},
				};

				it(`should render ${sectionName} hash link`, () => {
					render(<TreatmentSummaryPage {...props} />);

					expect(
						screen.queryByRole("link", { name: sectionName })
					).toHaveAttribute("href", `#${expectedId}`);
				});

				it(`should render ${sectionName} section with accessible name`, () => {
					render(<TreatmentSummaryPage {...props} />);

					expect(
						screen.getByRole("region", { name: sectionName })
					).toBeInTheDocument();
				});

				it(`should render list of ${sectionName} with accessible name`, () => {
					render(<TreatmentSummaryPage {...props} />);

					expect(
						screen.getByRole("list", { name: sectionName })
					).toBeInTheDocument();
				});

				it(`should render list item per related item`, () => {
					render(<TreatmentSummaryPage {...props} />);

					const relatedItemsList = screen.getByRole("list", {
						name: sectionName,
					});

					expect(
						within(relatedItemsList).getAllByRole("listitem")
					).toHaveLength(2);
				});

				it(`should render anchor per related item`, () => {
					render(<TreatmentSummaryPage {...props} />);

					const relatedItemsList = screen.getByRole("list", {
						name: sectionName,
					});

					expect(within(relatedItemsList).getAllByRole("link")).toHaveLength(2);
				});

				it(`should render related item title as link text`, () => {
					render(<TreatmentSummaryPage {...props} />);

					expect(
						screen.getByRole("link", { name: "Thing 1" })
					).toBeInTheDocument();
				});

				it(`should render related item title as link text`, () => {
					render(<TreatmentSummaryPage {...props} />);

					expect(screen.getByRole("link", { name: "Thing 1" })).toHaveAttribute(
						"href",
						`/${expectedBasePath}/thing-1/`
					);
				});
			});
		});
	});
});
