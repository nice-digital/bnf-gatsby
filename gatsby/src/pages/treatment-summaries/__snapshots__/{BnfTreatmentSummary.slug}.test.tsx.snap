// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreatmentSummaryPage should match snapshot for graphql query 1`] = `
"
	query ($id: String) {
		bnfTreatmentSummary(id: { eq: $id }) {
			title
			sections {
				...RecordSection
			}
		}
	}
"
`;
